cmd/libsnap-confine-private/device-cgroup-support.c

```diff
@@ -107,6 +111,355 @@ static void _sc_cgroup_v1_attach_pid(sc_device_cgroup *self, pid_t pid) {
+        sc_cgroup_v2_device_key key = {0};
+        size_t existing_count = 0;
+        while (true) {
+            sc_cgroup_v2_device_key next = {0};
+            if (existing_count >= max_entries) {
+                die("too many elements in the map");
+            }
+            if (existing_count > 0) {
+                /* grab the previous key */
+                key = existing_keys[existing_count - 1];
+            }
+            int ret = bpf_map_get_next_key(devmap_fd, &key, &next);
+            if (ret == -1) {
+                if (errno != ENOENT) {
+                    die("cannot lookup existing device map keys");
+                }
+                /* we are done */
+                break;
+            }
+            existing_keys[existing_count] = next;
+            existing_count++;
+        }
+        debug("found %zu existing entries in devices map", existing_count);
+        if (existing_count > 0) {
+#if 0
+            /* XXX: we should be doing a batch delete of elements, however:
+             * - on Arch with 5.13 kernel I'm getting EINVAL
+             * - the linux/bpf.h header present during build on 16.04 does not
+             *     support batch operations
+             */
+            if (bpf_map_delete_batch(devmap_fd, existing_keys, existing_count) < 0) {
+                 die("cannot dump all elements from devices map");
+            }
+#endif
+            for (size_t i = 0; i < existing_count; i++) {
+                sc_cgroup_v2_device_key key = existing_keys[i];
+                debug("delete key for %c %d:%d", key.type, key.major, key.minor);
+                if (bpf_map_delete_elem(devmap_fd, &key) < 0) {
+                    die("cannot delete device map entry for %c %d:%d", key.type, key.major, key.minor);
+                }
+            }
+        }
+    }
+
+    if (!from_existing) {
+        /* 4.5+ kernels allow maps to be created by simple users otherwise the
+         * calls fail with EPERM */
+        int prog_fd = load_devcgroup_prog(devmap_fd);
+        if (bpf_prog_attach(BPF_CGROUP_DEVICE, cgroup_fd, prog_fd) < 0) {
+            die("cannot attach cgroup program");
+        }
+    }
+
+    self->v2.devmap_fd = devmap_fd;
+    self->v2.cgroup_fd = cgroup_fd;
+
+    return 0;
+}
+
+static void _sc_cgroup_v2_close_bpf(sc_device_cgroup *self) {
+    sc_cleanup_string(&self->v2.tag);
+    /* the map is pinned to a per-snap-application file and referenced by the
+     * program */
+    sc_cleanup_close(&self->v2.devmap_fd);
+    sc_cleanup_close(&self->v2.cgroup_fd);
+}
+
+static void _sc_cgroup_v2_allow_bpf(sc_device_cgroup *self, int kind, int major, int minor) {
+    struct sc_cgroup_v2_device_key key = {
+        .major = major,
+        .minor = minor,
+        .type = (kind == S_IFCHR) ? 'c' : 'b',
+    };
+    sc_cgroup_v2_device_value value = 1;
+    debug("v2 allow %c %u:%u", (char)key.type, key.major, key.minor);
+    if (bpf_update_map(self->v2.devmap_fd, &key, &value) < 0) {
```

Comment:
does adding to a map need to be done by root? I'm a bit confused that we can create a map, delete items from it, and add items to it all by non-root, but the pinning to the path is the only part that needs to be done as root? Does that leave us open to the possibility of a non-root process (albeit unconfined) adding permissions to a confined snap and sort of doing privilege escalation here?

Reply:
AFAIU the current model enforced by the kernel is similar to open(), that is if a process can obtain a file descriptor to the object further operations are unrestricted.

Reply:
What is the current permissions of the path that we use for these? I see that it is created /sys/fs/bpf/snap is owned by root, but it's unclear to me if that's i.e. 0700 or something

Reply:
```
$ sudo find /sys/fs/bpf/ -ls
        1      0 drwx-----T   3 root     root            0 wrz 13 18:41 /sys/fs/bpf/
    27226      0 drwx------   2 root     root            0 wrz 13 18:45 /sys/fs/bpf/snap
    65984      0 -rw-------   1 root     root            0 wrz 13 18:45 /sys/fs/bpf/snap/snap_ohmygiraffe_ohmygiraffe
```

Reply:
Okay, so the paths are such that a normal user cannot just open this path and write to it. Thanks for clarifying that, when we have more spread testing of cgroupsv2 and are ready to turn it on fully, I think we should have this sort of check in a spread test